packages/@react-spectrum/dialog/src/AlertDialog.tsx

```diff
@@ -64,49 +64,60 @@ function AlertDialog(props: SpectrumAlertDialogProps, ref: DOMRef) {
     }
   }
 
+  let onKeyDown = (e) => {
```

Comment:
I looked at this PR and I feel it should be solved differently, but with my quick research I don't have a better solution. Maybe the following comments will convey a different direction. I expected the key event to be in react-aria, probably next to https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/overlays/src/useOverlay.ts#L112-L119. I'm not sure how onPrimaryAction would be called that way. Also, thinking it would be better to add to the existing Dialog context rather than creating a new one.

Reply:
@ktabors thanks for the feedback! I'll rethink the problem and (hopefully) come up with a better approach.

Reply:
So I think this context makes sense. It doesn't get in the way of the Dialog context and would technically sit between a DialogTrigger and a Dialog. So if we used the DialogContext, we'd overwrite that and have to merge it, but that would lead to other problems. These are both only Spectrum components, so also not worried about the extra context with that.

That said, I also tried this out using document level listeners on Enter key down, and it's definitely doable. Something along these lines

```
  let domRef = useDOMRef(null);
  let dialogRef: DOMRef = ref || domRef;
// ...
  let onKeyDown = (e) => {
    console.log(dialogRef, e.target);
    let dialogNode: HTMLElement = dialogRef.current?.UNSAFE_getDOMNode();
    if (e.key === 'Enter' && allowsKeyboardConfirmation && dialogNode === e.target) {
      e.stopPropagation();
      e.preventDefault();
      onPrimaryAction();
      onClose();
    }
  };

  let {addGlobalListener, removeAllGlobalListeners} = useGlobalListeners();
  useEffect(() => {
    addGlobalListener(document.body, 'keydown', onKeyDown);
    return () => {
      removeAllGlobalListeners();
    }
  }, []);
```

This also fixes an issue that I noticed, which is if you put anything that you would expect to hit the 'Enter' key on inside of the AlertDialog's content, then it would close the AlertDialog. To fix this, I am checking that the keydown event actually happened on the Dialog element itself, not on any of its children. That's the `dialogNode === e.target`
Note: TS for my example is probably failing.

I think it'd be good to add this story which demonstrates the faulty behavior. And we should base some tests off it as well.

```
//...
  .add(
    'alert with content actions',
    () => renderWithContentButtons({
      variant: 'error',
      title: 'Error: Danger Will Robinson',
      primaryActionLabel: 'Accept',
      cancelLabel: 'Cancel',
      onPrimaryAction: action('primary'),
      onSecondaryAction: action('secondary'),
      onCancel: action('cancel'),
      allowsKeyboardConfirmation: true
    })
  )
//...
function renderWithContentButtons(props) {
  return (
    <div style={{display: 'flex', width: 'auto', margin: '100px 0'}}>
      <DialogTrigger defaultOpen>
        <ActionButton>Trigger</ActionButton>
        <AlertDialog {...props} onPrimaryAction={action('primary')} onSecondaryAction={action('secondary')} onCancel={props.onCancel}>
          <Counter />
        </AlertDialog>
      </DialogTrigger>
    </div>
  );
}
function Counter() {
  const [count, setCount] = useState(0);

  const handleIncrement = () => {
    setCount(prevCount => prevCount + 1);
  };

  const handleDecrement = () => {
    setCount(prevCount => prevCount - 1);
  };
  return (
    <div>
      <div>
        <ActionButton onPress={handleDecrement}>-</ActionButton>
        <div>Count is {count}</div>
        <ActionButton onPress={handleIncrement}>+</ActionButton>
      </div>
      <ActionButton onPress={() => setCount(0)}>Reset</ActionButton>
    </div>
  );
}
```

Reply:
Thanks for the feedback @snowystinger ! I'll add this check, a story to demonstrate the faulty behavior, and some tests in my PR

Reply:
Hi Robert, I was trying to repro the issue you mentioned
`
if you put anything that you would expect to hit the 'Enter' key on inside of the AlertDialog's content, then it would close the AlertDialog.
`
I found that even though I don't have the check `dialogNode === e.target`, it's working as expected - only when the keydown event actually happens on the Dialog element itself will it closes the dialog, otherwise, it calls the children's own event handlers. In your example, if I set focus on the ActionButtons through the tab key and hit Enter, the increment/decrement/reset all work fine. So I'm wondering if I had some misunderstandings here. Any ideas?

Reply:
It's possible we now prevent some event through usePress that we weren't at the time of my comment. Can you try with a native button, no usePress?

Reply:
Oh yes, the issue can occur with native buttons, we still need that check. Thanks!

Reply:
Hi, I've added the following in the newest commit -
1. the check we talked about
2. a story to demonstrate the faulty use case
3. a test

I'm not sure if the way I handles ref makes sense though. Any feedbacks are welcome! Thanks in advance.